diff -Nur libsemanage-old/src/conf-parse.y libsemanage-2.0.43/src/conf-parse.y
--- libsemanage-old/src/conf-parse.y	2009-12-01 21:46:52.000000000 +0100
+++ libsemanage-2.0.43/src/conf-parse.y	2012-10-15 15:44:51.355161296 +0200
@@ -57,7 +57,7 @@
 }
 
 %token MODULE_STORE VERSION EXPAND_CHECK FILE_MODE SAVE_PREVIOUS SAVE_LINKED
-%token LOAD_POLICY_START SETFILES_START DISABLE_GENHOMEDIRCON HANDLE_UNKNOWN
+%token LOAD_POLICY_START SETFILES_START DISABLE_GENHOMEDIRCON HANDLE_UNKNOWN USEPASSWD
 %token BZIP_BLOCKSIZE BZIP_SMALL
 %token VERIFY_MOD_START VERIFY_LINKED_START VERIFY_KERNEL_START BLOCK_END
 %token PROG_PATH PROG_ARGS
@@ -82,6 +82,7 @@
         |       save_previous
         |       save_linked
         |       disable_genhomedircon
+        |		usepasswd
         |       handle_unknown
 	|	bzip_blocksize
 	|	bzip_small
@@ -153,6 +154,17 @@
 	free($3);
  }
 
+usepasswd: USEPASSWD '=' ARG {
+	if (strcasecmp($3, "false") == 0) {
+		current_conf->usepasswd = 0;
+	} else if (strcasecmp($3, "true") == 0) {
+		current_conf->usepasswd = 1;
+	} else {
+		yyerror("usepasswd can only be 'true' or 'false'");
+	}
+	free($3);
+}
+
 handle_unknown: HANDLE_UNKNOWN '=' ARG {
 	if (strcasecmp($3, "deny") == 0) {
 		current_conf->handle_unknown = SEPOL_DENY_UNKNOWN;
@@ -252,6 +264,7 @@
 	conf->policyvers = sepol_policy_kern_vers_max();
 	conf->expand_check = 1;
 	conf->handle_unknown = -1;
+	conf->usepasswd = 1;
 	conf->file_mode = 0644;
 	conf->bzip_blocksize = 9;
 	conf->bzip_small = 0;
diff -Nur libsemanage-old/src/conf-scan.l libsemanage-2.0.43/src/conf-scan.l
--- libsemanage-old/src/conf-scan.l	2009-12-01 21:46:51.000000000 +0100
+++ libsemanage-2.0.43/src/conf-scan.l	2012-10-15 15:36:08.653862592 +0200
@@ -46,6 +46,7 @@
 save-previous     return SAVE_PREVIOUS;
 save-linked       return SAVE_LINKED;
 disable-genhomedircon return DISABLE_GENHOMEDIRCON;
+usepasswd         return USEPASSWD;
 handle-unknown    return HANDLE_UNKNOWN;
 bzip-blocksize	return BZIP_BLOCKSIZE;
 bzip-small	return BZIP_SMALL;
diff -Nur libsemanage-old/src/direct_api.c libsemanage-2.0.43/src/direct_api.c
--- libsemanage-old/src/direct_api.c	2012-10-15 15:28:37.439958674 +0200
+++ libsemanage-2.0.43/src/direct_api.c	2012-10-15 15:36:54.885153793 +0200
@@ -970,7 +970,7 @@
 	 * which requires the out policydb */
 	if (!sh->conf->disable_genhomedircon) {
 		if (out && (retval =
-		     semanage_genhomedircon(sh, out, 1)) != 0) {
+		     semanage_genhomedircon(sh, out, sh->conf->usepasswd)) != 0) {
 			ERR(sh, "semanage_genhomedircon returned error code %d.",
 			    retval);
 			goto cleanup;
diff -Nur libsemanage-old/src/semanage_conf.h libsemanage-2.0.43/src/semanage_conf.h
--- libsemanage-old/src/semanage_conf.h	2009-12-01 21:46:51.000000000 +0100
+++ libsemanage-2.0.43/src/semanage_conf.h	2012-10-15 15:35:25.329589859 +0200
@@ -38,6 +38,7 @@
 	int save_previous;
 	int save_linked;
 	int disable_genhomedircon;
+	int usepasswd;
 	int handle_unknown;
 	mode_t file_mode;
 	int bzip_blocksize;
